Add guide for tuning kNN search

jtibshirani opened the pull request:
This 'how to' guide explains performance considerations specific to kNN search.
It takes inspiration from the 'tune for search speed' guide.

jtibshirani committed:
Add guide for tuning kNN search

This 'how to' guide explains performance considerations specific to kNN search.
It takes inspiration from the 'tune for search speed' guide.

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

abdonpijpelink commented on docs/reference/how-to/search-speed.asciidoc:
```diff
@@ -10,6 +10,7 @@ goes to the filesystem cache so that Elasticsearch can keep hot regions of the
 index in physical memory.
 
 [discrete]
+tag::readahead[]
```
```suggestion
tag::readahead[]
[discrete]
```
Is "Avoid page cache thrashing by using modest readahead values on Linux" broken out into its own page on purpose? If not, swap these two lines.

jtibshirani replied:
Oops, thanks!

abdonpijpelink reviewed (approved):
I've left one suggestion. LGTM otherwise

mayya-sharipova commented on docs/reference/how-to/knn-search.asciidoc:
```diff
@@ -0,0 +1,98 @@
+approximate kNN search. HNSW is a graph-based algorithm which only works
+efficiently when all vector data is held in memory. You should ensure that
+data nodes have enough RAM to hold the vector data and index structures. To
+check the size of the vector data, you can use the <<indices-disk-usage>> API.
```
Just to confirm, we don't use any compression when storing vector data on disk?

mayya-sharipova replied:
Should we provide an approximate formulate for the lower threshold for RAM: `N of docs * dims * 4` for vectors + `N of docs * M * 2` for graph?

jtibshirani replied:
Could you explain how compression is related here? And sounds good, I can add an approximate calculation here.

mayya-sharipova replied:
> Could you explain how compression is related here?

Sorry for a confusion, we are not using any disk compression for the approximate knn vector values.  We compress only binary doc values for the exact search.

jtibshirani replied:
I pushed an update to improve this section.

mayya-sharipova commented on docs/reference/how-to/knn-search.asciidoc:
```diff
@@ -0,0 +1,98 @@
+
+The `cosine` option accepts any float vector and computes the cosine
+similarity. While this is convenient for testing, it's not the most efficient
+approach. Instead, we recommend using the `dot_product` option for cosine
```
did you mean here "for vector similarity" instead of "cosine similarity"?

jtibshirani replied:
I did mean "cosine similarity", because `dot_product` is being used to compute the cosine similarity between the two vectors. I see how this is confusing though, I could just say "Instead, we recommend using the `dot_product` option."

mayya-sharipova commented on docs/reference/how-to/knn-search.asciidoc:
```diff
@@ -0,0 +1,98 @@
+downsides of omitting fields from `_source`.
+
+[discrete]
+=== Force merge to one segment
```
Should we say "Force merge to one segment for read-only indices"?

jtibshirani replied:
I discussed this with our team earlier, and we agreed that our guidance around force merging is too strict. I opened https://github.com/elastic/elasticsearch/pull/90151 to soften the restriction around force merging.

mayya-sharipova replied:
Thanks Julie, may be then, we should provide a link to this "force_merge" document, or add a warning here that force merge is recommended only for indices with few edits and deletes.

inqueue commented on docs/reference/how-to/knn-search.asciidoc:
```diff
@@ -0,0 +1,98 @@
+{es} shards are composed of segments, which are internal storage elements in the
+index. For approximate kNN search, {es} stores the dense vector values of each
+segment as an HNSW graph. <<indices-forcemerge,Force merging>> the index to a
+single segment can improve approximate kNN search latency. wWhen there are
```
s/`wWhen`/`When`/

jtibshirani committed:
Merge remote-tracking branch 'upstream/main' into knn-docs

jtibshirani committed:
Avoid breaking filesystem cache into its own page

jtibshirani committed:
Fix typo

jtibshirani committed:
Clarify dot_product section

jtibshirani commented:
Thanks for the reviews! There's one missing piece I'm still working on: what advice should we give to keep the number of segments as low as possible? Recently we've discussed several options:
* Force merge to one segment when possible
* Set `refresh_interval` to -1 to prevent a bunch of small segments from being created before the shard is marked "search idle"
* Use a larger heap size or increase the indexing memory buffer (?)

jtibshirani committed:
Merge remote-tracking branch 'upstream/main' into knn-docs

jtibshirani committed:
Improve the section on RAM usage

henningandersen reviewed (commented):
Left a few drive-by comments.

jtibshirani committed:
Fix typo

jtibshirani commented on docs/reference/how-to/knn-search.asciidoc:
```diff
@@ -0,0 +1,106 @@
+{es} shards are composed of segments, which are internal storage elements in the
+index. For approximate kNN search, {es} stores the dense vector values of each
+segment as an HNSW graph. <<indices-forcemerge,Force merging>> the index to a
+single segment can improve approximate kNN search latency. When there are
```
This is a good point! I'll think about what we could recommend here.

jtibshirani commented:
@mayya-sharipova @abdonpijpelink @inqueue this is ready for another review. Changes since last time:
* Created top-level section "Reduce the number of index segments", with subsections on force merge and increasing the indexing buffer.
* Improved the section on RAM usage by giving a rough estimate

inqueue reviewed (approved):
LGTM, thank you!

mayya-sharipova reviewed (approved):
Thanks @jtibshirani, new changes LGTM as well.
